Core/ALife.Core/WorldObjects/Agents/Brains/NeuralNetworkBrain.cs

```diff
@@ -28,217 +31,318 @@ public NeuralNetworkBrain(Agent self, List<int> layers)
-                }
+            //This makes the agent enact those items.
+            foreach(ActionCluster act in self.Actions.Values)
+            {
+                act.ActivateAction();
             }
+        }
 
-            Layer actionLayer = CreateClonedActionLayer(self, Layers[Layers.Count - 1], templateBrain.Layers[Layers.Count], exactCopy);
-            Layers.Add(actionLayer);
-            actions = actionLayer;
+        public IBrain Clone(Agent self)
+        {
+            return new NeuralNetworkBrain(self, this, true);
         }
 
-        private void CreateDendrites(List<Neuron> aboveLayerNeurons, Neuron templateNeuron, Neuron newNeuron, bool exactCopy)
+        public IBrain Reproduce(Agent self)
         {
-            for(int d = 0; d < aboveLayerNeurons.Count; d++)
-            {
-                double denWeight = templateNeuron.UpstreamDendrites[d].Weight;
-                if(!exactCopy)
-                {
-                    denWeight = ModifyBetweenNegOneAndOne(denWeight);
-                }
-                newNeuron.UpstreamDendrites.Add(new Dendrite(aboveLayerNeurons[d], denWeight));
-            }
+            return new NeuralNetworkBrain(self, this, false);
         }
 
-        private double ModifyBetweenNegOneAndOne(double original)
+        public string ExportNewBrain()
         {
-            double val = original;
-            double shouldMod = Planet.World.NumberGen.NextDouble();
-            if(shouldMod < ModificationRate)
+            //SPEC V2
+            //Line1: Bitstream
+            //Line2: NeuronNameArray
+
+            //bitstream
+            //int (number of layers) = L
+            //int[L] neurons in each layer, sum = N
+            //double ModificationRate
+            //double MutationRate
+            //double[N]  NeuronBiases
+            //L double[n*pn] DendriteWeights for the neuron and parentNeurons
+
+
+            byte layerCount = (byte)Layers.Count;
+            byte[] layerInfoArray = new byte[layerCount+1];
+            layerInfoArray[0] = layerCount;
+            for(int i = 0; i < layerCount; ++i)
             {
-                double rawMod = Planet.World.NumberGen.NextDouble();
-                double modification = ((rawMod * 2) - 1) * MutabilityRate;
-                val += modification;
-                val = ExtraMath.Clamp(val, -1, 1);
+                layerInfoArray[i + 1] = (byte)Layers[i].Neurons.Count;
             }
-            return val;
-        }
 
-        private Layer CreateActionLayer(Agent self, Layer aboveLayer)
-        {
-            List<ActionNeuron> actionNeurons = new List<ActionNeuron>();
-            foreach(ActionCluster ac in self.Actions.Values)
+            List<double> doubleInfo = new List<double>();
+            doubleInfo.Add(ModificationRate);
+            doubleInfo.Add(MutabilityRate);
+
+            List<double> biases = new List<double>();
+            for(int j = 0; j < layerCount; ++j)
```

Comment:
Let's add a comment to at a later that:
- Add directly to the doubleInfo array
- Either:
  - (preferred) hide each for loop into a separate method to clean things up
  - Merge the for loop into one

Reply:
Added some comments.

These cannot be re-ordered, because the double array is strictly ordered. Moving anything would break it.